File: cpp/Ch 7.OOD/1.Deck Of Cards/BlackJackHand.cpp
#include <vector>
#include <limits>

class BlackJackHand : public Hand<BlackJackCard*>
{
public:
	BlackJackHand()
	{

	}

	virtual int score()
	{
		std::vector<int> scores = possibleScores();
		int maxUnder = std::numeric_limits<int>::min();
		int minOver = std::numeric_limits<int>::max();
		for (auto score : scores)
		{
			if (score > 21 && score < minOver)
			{
				minOver = score;
			}
			else if (score <= 21 && score > maxUnder)
			{
				maxUnder = score;
			}
		}
		return maxUnder == std::numeric_limits<int>::min() ? minOver : maxUnder;
	}

private:
	void std::vector<int> possibleScores()
	{
		std::vector<int> scores;
		if (cards->size() == 0)
		{
			return scores;
		}
		for (BlackJackCard *card : cards)
		{
			addCardToScoreList(card, scores);
		}
		return scores;
	}

	void addCardToScoreList(BlackJackCard *card, std::vector<int> &scores)
	{
		if (scores.empty())
		{
			scores.push_back(0);
		}
		int length = scores.size();
		for (int i = 0; i < length; i++)
		{
			int score = scores[i];
			scores[i] = score + card->minValue();
			if (card->minValue() != card->maxValue())
			{
				scores.push_back(score + card->maxValue());
			}
		}
	}

public:
	virtual bool busted()
	{
		return score() > 21;
	}

	virtual bool is21()
	{
		return score() == 21;
	}

	virtual bool isBlackJack()
	{
		if (cards->size() != 2)
		{
			return false;
		}
		BlackJackCard *first = cards->get(0);
		BlackJackCard *second = cards->get(1);
		return (first->isAce() && second->isFaceCard()) || (second->isAce() && first->isFaceCard());
	}
};

File: Java/Ch 07. Object-Oriented Design/Q7_01_Deck_of_Cards/BlackJackHand.java
package Q7_01_Deck_of_Cards;

import java.util.ArrayList;

public class BlackJackHand extends Hand<BlackJackCard> {
	public BlackJackHand() {
		
	}
	
	public int score() {
		ArrayList<Integer> scores = possibleScores();
		int maxUnder = Integer.MIN_VALUE;
		int minOver = Integer.MAX_VALUE;
		for (int score : scores) {
			if (score > 21 && score < minOver) {
				minOver = score;
			} else if (score <= 21 && score > maxUnder) {
				maxUnder = score;
			}
		}
		return maxUnder == Integer.MIN_VALUE ? minOver : maxUnder;
	}
	
	private ArrayList<Integer> possibleScores() {
		ArrayList<Integer> scores = new ArrayList<Integer>();
		if (cards.size() == 0) {
			return scores;
		}
		for (BlackJackCard card : cards) {
			addCardToScoreList(card, scores);
		}
		return scores;
	}
	
	private void addCardToScoreList(BlackJackCard card, ArrayList<Integer> scores) {
		if (scores.size() == 0) {
			scores.add(0);
		} 
		int length = scores.size();
		for (int i = 0; i < length; i++) {
			int score = scores.get(i);
			scores.set(i, score + card.minValue());
			if (card.minValue() != card.maxValue()) {
				scores.add(score + card.maxValue());
			}
		}
	}
	
	public boolean busted() {
		return score() > 21;
	}
	
	public boolean is21() {
		return score() == 21;
	}
	
	public boolean isBlackJack() {
		if (cards.size() != 2) {
			return false;
		}
		BlackJackCard first = cards.get(0);
		BlackJackCard second = cards.get(1);
		return (first.isAce() && second.isFaceCard()) || (second.isAce() && first.isFaceCard());
	}
}
